lib/Mojo/mysql/Connection.pm

```diff
@@ -0,0 +1,797 @@
+    $self->{seq}++;
+    $self->$next();
+  }
+  else {
+    $self->$next();
+  }
+}
+
+sub _seq {
+  my ($self, $cmd, $cb) = @_;
+
+  $self->{iostream} = Mojo::IOLoop::Stream->new($self->{socket});
+  $self->{iostream}->reactor($self->_ioloop(0)->reactor) unless $cb;
+  $self->{iostream}->timeout($self->options->{query_timeout})
+    if $self->options->{query_timeout};
+  weaken $self;
+
+  $self->{iostream}->on(read => sub {
+    my ($stream, $bytes) = @_;
+    $self->{incoming} .= $bytes;
+
+    $self->_seq_next($cmd, 0) while $self->_seq_next_ready;
+
+    if ($self->_state eq 'idle' or $self->_state eq 'error') {
+      $stream->steal_handle;
+      delete $self->{iostream};
+      $cb ? $self->$cb() : $self->_ioloop(0)->stop;
+    }
+    else {
+      $self->_seq_next($cmd, 1);
+    }
+  });
+  $self->{iostream}->on(error => sub {
+    my ($stream, $err) = @_;
+    warn "stream error: $err\n" if DEBUG;
+    $self->{error_message} //= $err;
+  });
+  $self->{iostream}->on(timeout => sub {
+    warn "stream timeout\n" if DEBUG;
+    $self->{error_message} //= 'timeout';
+  });
+  $self->{iostream}->on(close => sub {
+    $self->{socket} = undef;
+    $self->_state('disconnected');
+    $cb ? $self->$cb() : $self->_ioloop(0)->stop;
+  });
+
+  $self->_seq_next($cmd, 1);
+  $self->{iostream}->start;
+}
+
+sub _cmd {
+  my ($self, $cmd, $cb) = @_;
+  die 'invalid cmd:' . $cmd unless exists SEQ->{$cmd};
+  die 'invalid state:' . $self->_state . ' doing:'. $cmd unless exists SEQ->{$cmd}{$self->_state};
+
+  $self->_reset;
+  $self->_seq($cmd, $cb);
+  $self->_ioloop(0)->start unless $cb;
+  return $self->_state eq 'idle' ? 1 : 0;
+}
+
+sub connect {
+  my ($self, $cb) = @_;
+
+  $self->_state('connecting');
+  $self->_reset;
+
+  $self->{client} = Mojo::IOLoop::Client->new;
+  $self->{client}->reactor($self->_ioloop(0)->reactor) unless $cb;
+  weaken $self;
+
+  $self->{client}->on(connect => sub {
+    my ($client, $handle) = @_;
+    delete $self->{client};
+    $self->{socket} = $handle;
+    $self->_state('connected');
+    $self->_seq('connect', $cb);
+  });
+  $self->{client}->on(error => sub {
+    my ($client, $err) = @_;
+    delete $self->{client};
+    $self->_state('disconnected');
+    $cb ? $self->$cb() : $self->_ioloop(0)->stop;
+  });
+
+  $self->{client}->connect(
+    address => $self->host, port => $self->port,
+    timeout => $self->options->{connect_timeout}
+  );
+
+  $self->_ioloop(0)->start unless $cb;
+  die $self->{error_message} if $self->{error_code};
+}
+
+sub disconnect { shift->_cmd('disconnect') }
+
+sub query {
+  my ($self, $sql, $cb) = @_;
+  $self->{sql} = $sql;
+  $self->_cmd('query', $cb);
+}
+
+sub ping {
+  my $self = shift;
+  return $self->_state eq 'disconnected' ? 0 : $self->_cmd('ping');
+}
+
+sub DESTROY {
+  my $self = shift;
+  $self->unsubscribe($_) for qw(connect fields result end errors);
+  $self->disconnect if $self->_state eq 'idle' and $self->{socket};
```

Comment:
**unsubscribe** is normally not needed. Only if the application using it forgets to unsubscribe.
**disconnect** is a packet sequence to inform the server that client requests to end session.